File: domain/src/main/kotlin/com/griffgym/domain/model/Rpe.kt
package com.griffgym.domain.model

import kotlin.math.abs
import kotlin.math.roundToLong

/**
 * Rate of Perceived Exertion — subjective intensity of a single set, 1.0 .. 10.0 in 0.5 steps.
 */
@JvmInline
value class Rpe private constructor(val value: Double) : Comparable<Rpe> {

    override fun compareTo(other: Rpe): Int = value.compareTo(other.value)

    override fun toString(): String = format()

    fun format(): String =
        if (abs(value % 1.0) < EPSILON) value.toLong().toString() else value.toString()

    fun stepUp(): Rpe = ofOrNull(value + STEP) ?: this

    fun stepDown(): Rpe = ofOrNull(value - STEP) ?: this

    companion object {
        private const val EPSILON = 1e-6

        const val MIN_VALUE: Double = 1.0
        const val MAX_VALUE: Double = 10.0
        const val STEP: Double = 0.5

        val MIN: Rpe = Rpe(MIN_VALUE)
        val MAX: Rpe = Rpe(MAX_VALUE)

        fun of(value: Double): Rpe = requireNotNull(ofOrNull(value)) {
            "RPE must be within $MIN_VALUE..$MAX_VALUE in steps of $STEP, was $value"
        }

        /** Returns `null` instead of throwing — used for user input validation. */
        fun ofOrNull(value: Double): Rpe? {
            if (!value.isFinite()) return null
            val snapped = snap(value)
            if (snapped < MIN_VALUE || snapped > MAX_VALUE) return null
            return Rpe(snapped)
        }

        fun parse(raw: String): Rpe? {
            val normalised = raw.trim().replace(',', '.')
            if (normalised.isEmpty()) return null
            return normalised.toDoubleOrNull()?.let(::ofOrNull)
        }

        fun isValid(value: Double): Boolean = ofOrNull(value) != null

        /** Snaps to the nearest 0.5 so the UI can never persist an RPE of 7.31. */
        private fun snap(value: Double): Double = (value / STEP).roundToLong() * STEP
    }
}

/**
 * A planned intensity, either an exact value ("RPE 8") or a range ("RPE 6-7") as used
 * for accessory work.
 */
data class RpeTarget(val min: Rpe, val max: Rpe) {

    init {
        require(min <= max) { "RPE target range is inverted: $min > $max" }
    }

    val isRange: Boolean get() = min != max

    fun format(): String = if (isRange) "${min.format()}-${max.format()}" else min.format()

    companion object {
        fun exact(value: Double): RpeTarget = Rpe.of(value).let { RpeTarget(it, it) }

        fun range(min: Double, max: Double): RpeTarget = RpeTarget(Rpe.of(min), Rpe.of(max))
    }
}

File: GriffGym.Backend/src/GriffGym.Domain/Training/Rpe.cs
using GriffGym.Domain.Common;

namespace GriffGym.Domain.Training;

/// <summary>
/// Rate of Perceived Exertion: the subjective intensity of one set, 1.0 to 10.0 in half steps.
///
/// The half-step rule is enforced rather than rounded silently. A client that sends 7.31 has
/// a bug, and quietly storing 7.5 would hide it.
/// </summary>
public readonly record struct Rpe : IComparable<Rpe>
{
    public const decimal MinValue = 1.0m;
    public const decimal MaxValue = 10.0m;
    public const decimal Step = 0.5m;

    private Rpe(decimal value) => Value = value;

    public decimal Value { get; }

    public static Rpe Min { get; } = new(MinValue);

    public static Rpe Max { get; } = new(MaxValue);

    public static Rpe Of(decimal value)
    {
        DomainException.Require(
            IsValid(value),
            $"RPE must be within {MinValue}..{MaxValue} in steps of {Step}, was {value}.");

        return new Rpe(decimal.Round(value, 1, MidpointRounding.AwayFromZero));
    }

    public static Rpe? OfNullable(decimal? value) => value is null ? null : Of(value.Value);

    public static bool IsValid(decimal value) =>
        value >= MinValue && value <= MaxValue && value % Step == 0m;

    public int CompareTo(Rpe other) => Value.CompareTo(other.Value);

    public static bool operator <(Rpe left, Rpe right) => left.CompareTo(right) < 0;

    public static bool operator >(Rpe left, Rpe right) => left.CompareTo(right) > 0;

    public static bool operator <=(Rpe left, Rpe right) => left.CompareTo(right) <= 0;

    public static bool operator >=(Rpe left, Rpe right) => left.CompareTo(right) >= 0;

    public override string ToString() =>
        Value == decimal.Truncate(Value)
            ? decimal.Truncate(Value).ToString(System.Globalization.CultureInfo.InvariantCulture)
            : Value.ToString("0.#", System.Globalization.CultureInfo.InvariantCulture);
}

/// <summary>
/// A planned intensity: either exact ("RPE 8") or a range ("RPE 6-7"), as accessory work is
/// prescribed. Stored as two columns so a range never has to be parsed back out of a string.
/// </summary>
public readonly record struct RpeTarget
{
    private RpeTarget(Rpe min, Rpe max)
    {
        Min = min;
        Max = max;
    }

    public Rpe Min { get; }

    public Rpe Max { get; }

    public bool IsRange => Min != Max;

    public static RpeTarget Exact(decimal value) => Of(Rpe.Of(value), Rpe.Of(value));

    public static RpeTarget Range(decimal min, decimal max) => Of(Rpe.Of(min), Rpe.Of(max));

    public static RpeTarget Of(Rpe min, Rpe max)
    {
        DomainException.Require(min <= max, $"RPE target range is inverted: {min} > {max}.");
        return new RpeTarget(min, max);
    }

    /// <summary>
    /// Rebuilds a target from two nullable columns. Both null means "no target"; exactly one
    /// null is a corrupt row and says so rather than inventing the missing bound.
    /// </summary>
    public static RpeTarget? FromBounds(decimal? min, decimal? max)
    {
        if (min is null && max is null)
        {
            return null;
        }

        DomainException.Require(
            min is not null && max is not null,
            "An RPE target needs both bounds or neither.");

        return Of(Rpe.Of(min!.Value), Rpe.Of(max!.Value));
    }

    public override string ToString() => IsRange ? $"{Min}-{Max}" : Min.ToString();
}
